### ISSUE TITLE
Adding a Custom Component

### ISSUE THREAD

**Issue by TheFrankman** [saleem-hadad/larecipe#105]
**Describe the bug**
This is not actually a bug but neither is it a feature request.
I'm trying to add my own component and then rebuild the package.
However, as soon as a `yarn run dev` or `yarn run production` i get 404s for some of the assets 

myurl.local/vendor/binarytorch/larecipe/assets/css/dark.css
myurl.local/vendor/binarytorch/larecipe/assets/css/light.css

Do you have any documentation for how to add custom components and re-build the package (my components will be commited to my fork)

**To Reproduce**
Add a component TestComponent.vue with just `hello world` in template `name: test-component`

```
<template>
    <div>
        Hello World
    </div>
</template>

<script>
    export default {
        name : "test-component"
    }
</script>
```
Add new component to globalComponents.js

```
...
import TestComponent from "../components/TestComponent"
...
Vue.component(TestComponent.name, TestComponent);
```
Add `<test-component></test-component>` to overview.md
`yarn run production` or `yarn run dev`

Perhaps there is a way of doing this in the vendor directory, but i'm not sure how. 

**Expected behavior**
I would see Hello World in the overview section of the docs but I don't also, the 404s as mentioned above. What am i missing here ?

**Comment by TheFrankman** [comment]
Ok, so I made a little progress here. 
Currently i'm modifying the files directly in vendor (later i'll copy the changes over to my fork)

- From the vendor directory, `yarn run dev` or `yarn run production`
- Delete laravelroot/public/vendor/binarytorch/larecipe/assets/js/app.js
- Run php artisan larecipe:install

This re-publishes the app.js file in public directory and the new component will work.

However, my original question stands, is there an easier way to create custom components for la-recipe, or can this be added as a feature request as i'm sure it would be a widely adopted feature.

**Comment by saleem-hadad** [comment]
Hi @TheFrankman 
Thanks for your time
Actually adding new components (totally different than the provided ones) is really easy 
all you have to do is to compile your components into a js file under `public` folder
then in the `config/larecipe.php` file add the path to that js file in the additional JS/CSS section
I suggest you use Laravel mix to help you do the bundling stuff

**Comment by saleem-hadad** [comment]
let me if this is clear or should I investigate more on this

**Comment by TheFrankman** [comment]
Hello @saleem-hadad
Thanks for the quick response. I'll look into this today / tomorrow and get back to you.
But this makes sense, so i'm sure there wont be any issues.
It would be good if this could be documented, I was really surprised to see there were no closed issues regarding the same question, thought I must have been missing something obvious.

**Comment by TheFrankman** [comment]
Hello @saleem-hadad 

I have attempted to do this and my components are not being loaded (whereas they do work if i add them to the packed itself). I'll do my best to demonstrate what i've done, i've done by best emulate exactly what you do in your package : 

Laravel Project`webpack.mix.js`

```
mix
  .js('resources/assets/js/docs/all.js', 'public/assets/platform/js/docs')
  .extract()
```

`resources/assets/js/docs/all.js`
```
import globalComponents from "./globalComponents";

export default {
    install(Vue) {
        Vue.use(globalComponents);
    }
};
```

`resources/assets/js/docs/globalComponents.js`
```
import AccordionItem from "VueComponents/Misc/Accordions/AccordionItem"
import CodeExamples from "VueComponents/Docs/CodeExamples";
import JsonResponse from "VueComponents/Docs/JsonResponse";

export default {
    install(Vue) {
        Vue.component('accordion-item', AccordionItem);
        Vue.component(CodeExamples.name, CodeExamples);
        Vue.component(JsonResponse.name, JsonResponse);
    },
}
```

`config/larecipe.php`

```
'additional_js' => [
    '/assets/platform/docs/js/manifest.js',
    '/assets/platform/docs/js/vendor.js',
    '/assets/platform/docs/js/all.js'
],
```

It's important to note that there aren no 404s or anything as obvious as that. 

What am I missing here ?

**Comment by saleem-hadad** [comment]
Hello @TheFrankman 
Thanks for your time explaining the process
Unfortunately, you're correct, at this moment it's not possible to add custom vue components with the current structure of Larecipe.
However, now I'm working on changing the internal structure of larecipe with plugin support so that you can easily add new components within seconds, hopefully, this will be available at the beginning of the next month.

**Comment by TheFrankman** [comment]
Hello @saleem-hadad 
That would be great, It will obviously make Larecipe a lot better, you're never going to be able to add everything that everyone needs, so making components as easy to add as possible will save a lot of future development.

**Comment by TheFrankman** [comment]
@saleem-hadad 
I know it's only the 1st of April, but what kind of timeframe are we looking at for the new internal structure ? 

Let me know if there is anything you'd like me to test.

**Comment by saleem-hadad** [comment]
Hi @TheFrankman 
Thanks for the reminder, actually yesterday I published the new version of Larecipe v2.0 which contains the new structure so that you can add your own custom components very easily.
I've updated the docs yesterday on how to create custom assets(components), [see here](https://larecipe.binarytorch.com.my/docs/1.3/custom-css-and-js) and if you have any question I'll try to help you 😄

**Comment by TheFrankman** [comment]
@saleem-hadad - Sorry for not checking for updates before bothering you. Facepalm. I'll check it out and i'll let you know how it goes!

**Comment by saleem-hadad** [comment]
no issues, glad to help 😎

**Comment by TheFrankman** [comment]
Hello @saleem-hadad 

I'm struggling a bit here, likely the documentation could just do with a bit more detail.

I have run `php artisan larecipe:asset your-vendor/asset-name` (replacing the names for my needs)

That has created the `larecipe-components/MyAsset` directory as expected. Within that, i've added by two new components "JsonResponse" & "Code Examples"

asset.js then looks like this : 

```
import CodeExamples from './CodeExamples';
import JsonResponse from './JsonResponse';

LaRecipe.booting((Vue) => {
    Vue.component(CodeExamples.name, CodeExamples);
    Vue.component(JsonResponse.name, JsonResponse);
})
```

I then run `npm run dev` and my components don't work in the docs

The additional things I have done

- Upgraded to v2 following the steps on the upgrade section (frontend all works)
- Deleted `public/vendor/binarytorch/larecipe/assets/js/app.js`
- Re-run `php artisan larecipe:install`
- Cleared browser & laravel caches

**Comment by saleem-hadad** [comment]
Hi @TheFrankman 
Thanks for reporting
Where did you run this command `npm run dev`? you should be inside the package `larecipe-components/MyAsset` and not inside your Laravel project
also make sure you use the correct name for components as in your docs

**Comment by saleem-hadad** [comment]
Also, can I see the error message?

**Comment by saleem-hadad** [comment]
I've tried it now and it seems everything is working.
This is the components that come with a new asset package as a demo
![Screenshot 2019-04-01 at 11 10 57 PM](https://user-images.githubusercontent.com/12780866/55338510-72101500-54d3-11e9-8101-264d18ebd48d.png)

**Comment by saleem-hadad** [comment]
and here's another one

<img width="898" alt="Screenshot 2019-04-01 at 11 13 48 PM" src="https://user-images.githubusercontent.com/12780866/55338722-e21e9b00-54d3-11e9-9e10-1cda28e58bc6.png">
<img width="723" alt="Screenshot 2019-04-01 at 11 13 59 PM" src="https://user-images.githubusercontent.com/12780866/55338725-e480f500-54d3-11e9-983b-d99e22e2adbe.png">
![Screenshot 2019-04-01 at 11 14 02 PM](https://user-images.githubusercontent.com/12780866/55338732-e77be580-54d3-11e9-8570-22f1219f056c.png)

**Comment by TheFrankman** [comment]
Hello @saleem-hadad 

I think i know where I went wrong.

Now that I have done some digging and seen the changes that are made to the composer.json in the project root I understand what you are up to! Very clever, I like the approach.

During the setup stages of `php artisan larecipe:asset` I said 'no' when it asked if I wanted to update composer packages because I didn't initially see the point.

I started fresh, adding my components, then ran `npm run dev` and everything has now worked!
